demos-europe/demosplan-core
# F dplan 11483 additional permissions paid orga

### Ticket
https://demoseurope.youtrack.cloud/issue/DPLAN-11483/8-2-Als-Planungsburo-brauche-ich-eine-Admin-Rolle-um-Verfahren-selbst-anlegen-und-durchfuhren-zu-konnen.

### Description
**Goal** Allow for this case and for the future to add permissions based on certain conditions. In this case, add the permissions 'feature_admin_new_procedure' to be allowed for certain orga, role and customer


**Step 1** With SUPPORT role, update orga: click on checkbox "Darf Verfahren anlegen": 
- If checked, then a new permission (feature_admin_new_procedure) is created with the following criteria: user role PRIVATE_PLANNING_AGENCY, the current orga id and the current customer id
- The SUPPORT role also has a new permission 'feature_orga_new_procedure_edit' that enables this checkbox "Darf Verfahren anlegen" to be displayed.

**Technical details:** 
a) New Entity demosplan/DemosPlanCoreBundle/Logic/Permission/AccessControlPermissionService.php was created with index: permission, role,customer,orga so that it is not possible to add repeated permissions
b) New Service demosplan/DemosPlanCoreBundle/Logic/Permission/AccessControlPermissionService.php was created. These are the ground work for permissions. Some examples:
- If permission has orga, role, customer not null (like the case of this pr) then the permission is given for the orga, role, customer combination
- if permission has role and customer not null, and orga null: then the permission is given for all orgas on that particular customer and role
- and so on. That said: null is a wildcard
- c) Only users with Support Role and permission 'feature_orga_new_procedure_edit' are able to click this chechbox. See demosplan/DemosPlanCoreBundle/Controller/User/DemosPlanOrganisationAPIController.php and demosplan/DemosPlanCoreBundle/ResourceTypes/OrgaResourceType.php


**Step 2** With  PRIVATE_PLANNING_AGENCY role in a Planing Office is logged-in on the updated orga/customer (of step 1), the user will be able to create a Procedure and edit it.

**Technical details:**
a) When a procedure is created, by default the PlaningOffices and AuthorizedUsers are copied from the Blueprint. We cannot do that in this case, as the orga that is creating the procedure should be the only one that can access it. This pr sets the PlaningOffices as the current orga as long as the user has the permission feature_admin_new_procedure. See this file demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php that 
b) If the Support Role of step 1 removes the permission of the orga to creare Procedures, the Orga should still be able to keep editing the created Procedures (that is why it is important to do a) point.

### How to review/test

1. Pick a user of an orga that has Fachplanung-Planungsbüro
2. . Logged in and check that you cannot create Procedures
3.  Logged In as support role
4. Update any field of the orga that is Planungs buro
5. Then logged in as the same user as step 1
6. Create a procedure
7. Go back to support role and remove the permission
8. Go back to the user role Fachplanung-Planungsbüro. You should be able to still edit your procedures and not create new ones

Delete the checkbox if it doesn't apply/isn't necessary.
